### PR TITLE
Fixes #24728 - Support Huawei VRP

### PR DESCRIPTION
**Pull request by UXabre**
Allow Huawei switches to be updates and configured via foreman

QUIRK: Only supports tokenless built URL

### CODE REVIEW

File: lib/net/dhcp/ztp_record.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+      self.firmware    = opts[:firmware]
+      self.vendor      = opts[:vendor]
+    end
+
+    def attrs
+      super.merge(
+        { 
+          :ztp_firmware   => firmware,
+          :ztp_vendor     => vendor
+        }).delete_if { |k, v| v.nil? }
```
Comment (timogoebel) [inline lib/net/dhcp/ztp_record.rb]: You should be able to use `compact` here.

File: app/models/concerns/orchestration/dhcp.rb [hunk 2]
```diff
@@ -96,7 +96,15 @@ def boot_server
   def build_dhcp_record(record_mac)
     raise ::Foreman::Exception.new(N_("DHCP not supported for this NIC")) unless dhcp?
     record_attrs = dhcp_attrs(record_mac)
-    record_type = (provision? && jumpstart?) ? Net::DHCP::SparcRecord : Net::DHCP::Record
+    record_type = Net::DHCP::Record
+
+    if provision?
```
Comment (timogoebel) [inline app/models/concerns/orchestration/dhcp.rb]: I would extract that to a method or service class.

Comment (lzap) [reply app/models/concerns/orchestration/dhcp.rb]: Probably with switch-case for now.

File: lib/net/dhcp/ztp_record.rb (hunk 1 above)
Comment (UXabre) [reply lib/net/dhcp/ztp_record.rb]: Good one, thank you!

File: app/views/unattended/provisioning_templates/ZTP/vrp_default_ztp_config.erb [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+user-interface vty 16 20
+#
+nqa test-instance admin foreman
+ test-type <%= foreman_url('built')[/(.*):\/\//, 1] || 'http' %>
+ destination-address url <%= foreman_url('built')[/:\/\/([a-z0-9\-._~%]+|\[[a-z0-9\-._~%!$&'()*+,;=:]+\])/,1] %>
+ destination-port <%= foreman_url('built')[/:([0-9]+)\//,1] || 8000 %>
+ http-url <%= foreman_url('built') %> v1.1
+ start now
+#
+return
```
Comment (lzap) [inline app/views/unattended/provisioning_templates/ZTP/vrp_default_ztp_config.erb]: Mind adding the missing newline?

File: test/fixtures/architectures.yml [hunk 4]
```diff
@@ -7,3 +7,6 @@ sparc:
 
 s390:
   name: s390
+
+ASIC:
+  name: ASIC
```
Comment (lzap) [inline test/fixtures/architectures.yml]: Ditto.

File: test/fixtures/media.yml [hunk 5]
```diff
@@ -21,3 +21,7 @@ solaris10:
 opensuse:
   name: OpenSUSE
   path: http://download.opensuse.org/distribution/$version/repo/oss
+
+vrp5:
+  name: Huawei VRP 5
+  path: http://firmware.fake-huawei.com
```
Comment (lzap) [inline test/fixtures/media.yml]: Here and couple of others.

File: app/models/concerns/orchestration/dhcp.rb (hunk 2 above)
Comment (UXabre) [reply app/models/concerns/orchestration/dhcp.rb]: Hi @timogoebel ,
Would something like this be okay for you?
```
  def get_record_type()
    record_type = Net::DHCP::Record
    if provision?
      if jumpstart?
        record_type = Net::DHCP::SparcRecord
      elsif operatingsystem.respond_to?(:pxe_type) && operatingsystem.pxe_type == "ZTP" && operatingsystem.respond_to?(:ztp_arguments)
        record_type = Net::DHCP::ZTPRecord
      end
    end
  end

  def build_dhcp_record(record_mac)
    raise ::Foreman::Exception.new(N_("DHCP not supported for this NIC")) unless dhcp?
    record_attrs = dhcp_attrs(record_mac)
    record_type = get_record_type

    handle_validation_errors do
      record_type.new(record_attrs)
    end
  end
```

@lzap I'm not sure if I understand, as I don't immediately see how I would convert this to a switch case? Or is it specifically for the 'pxe_type == "ZTP"' part on line 104? :-)

Comment (timogoebel) [reply app/models/concerns/orchestration/dhcp.rb]: Something like this:

```ruby
  def get_record_type
    if provision?
      if jumpstart?
        return Net::DHCP::SparcRecord
      elsif operatingsystem.respond_to?(:pxe_type) && operatingsystem.pxe_type == "ZTP" && operatingsystem.respond_to?(:ztp_arguments)
        return Net::DHCP::ZTPRecord
      end
    end
    Net::DHCP::Record
  end
```

Comment (timogoebel) [reply app/models/concerns/orchestration/dhcp.rb]: On second thought: Even better would be to implement a `dhcp_record_type` method on the OS, `operatingsystem.dhcp_record_type`.

Something like this (needs more work, just to show the idea):
In base:

```ruby
def dhcp_record_type
  Net::DHCP::Record
end
```

For Sparc:
```ruby
def dhcp_record_type
  Net::DHCP:: SparcRecord
end
```

For ZTP:

```ruby
def dhcp_record_type
  return Net::DHCP::ZTPRecord if pxe_type == "ZTP"
  super
end
```

Comment (UXabre) [reply app/models/concerns/orchestration/dhcp.rb]: Okay, will implement the latter, looks cleaner to me :-)

File: app/models/concerns/orchestration/tftp.rb [hunk 6]
```diff
@@ -69,11 +69,14 @@ def build_pxe_render(kind)
   end
 
   def default_pxe_render(kind)
-    template = ProvisioningTemplate.find_by_name(host.local_boot_template_name(kind))
-    raise Foreman::Exception.new(N_("Template '%s' was not found"), template.name) unless template
+    template_name = host.local_boot_template_name(kind)
+    # Safely return in case there's no template configured for the specified kind
+    return unless template_name.present?
```
Comment (lzap) [inline app/models/concerns/orchestration/tftp.rb:74]: Probably write a unit test for this.

File: app/models/operatingsystems/junos.rb [hunk 7]
```diff
@@ -25,7 +25,7 @@ def pxedir
   end
 
   def boot_filename(host = nil)
-    "ztp.cfg/" + host.mac.delete(':').upcase
+    "ztp.cfg/" + host.mac.delete(':').upcase + ".cfg"
```
Comment (lzap) [inline app/models/operatingsystems/junos.rb]: Please explain. JunOS is something different than Huawei thing, is it? Is this related? And why are you changing this? Note this is a breaking change.

Comment (UXabre) [reply app/models/operatingsystems/junos.rb]: Hey @lzap , in details it's different but they share similarities. In my smart-proxy patch I've also added the .cfg so it works. JunOS does not seem too picky on the name, however, Huawei VRP seems to not consider the configuration file if this extension is missing (as VRP supports other means as well). It didn't look so dangerous to add the extension; however, if you prefer, I can add the extension only if it's a huawei switch :-)

File: app/models/concerns/orchestration/tftp.rb (hunk 6 above)
Comment (UXabre) [reply app/models/concerns/orchestration/tftp.rb:74]: Good one, will do so :-)

File: app/models/operatingsystems/junos.rb (hunk 7 above)
Comment (lzap) [reply app/models/operatingsystems/junos.rb]: That's fine. Thanks.

File: app/models/concerns/orchestration/dhcp.rb (hunk 2 above)
Comment (lzap) [reply app/models/concerns/orchestration/dhcp.rb]: Yup good idea.

File: app/views/unattended/provisioning_templates/ZTP/vrp_default_ztp_config.erb (hunk 3 above)
Comment (lzap) [inline app/views/unattended/provisioning_templates/ZTP/vrp_default_ztp_config.erb]: Lemme take a look --- no I don't have any Huawei switches at home. ;-) I will trust you with this one, will just test rendering and deployment. Please file a PR into community-templates with this one.

If you want to make sure this doesn't break in the future, add this into `snapshots.yaml` and create a template snapshot via `bundle exec rake snapshots:generate RAILS_ENV=test`

**Review by lzap — DISMISSED** [review]
LGTM @timogoebel any comments? Then I can test.

**Review by timogoebel — DISMISSED** [review]
Code looks good.

File: app/models/operatingsystems/junos.rb (hunk 7 above)
Comment (UXabre) [reply app/models/operatingsystems/junos.rb]: Hey @lzap, just to be sure, you want me to add the code which only adds a .cfg for huawei switches (and thus remove the general addition of this extension for all ZTP switches), right? :-)

File: app/views/unattended/provisioning_templates/ZTP/vrp_default_ztp_config.erb (hunk 3 above)
Comment (UXabre) [reply app/views/unattended/provisioning_templates/ZTP/vrp_default_ztp_config.erb]: Okay I will move this template to community templates. I'll check out the second part of your comment, wasn't aware of this :-) thanks!

File: app/models/operatingsystems/vrp.rb [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    "ztp.cfg/images/#{medium_provider.unique_id}/"
+  end
+
+  # The PXE type to use when generating actions and evaluating attributes. jumpstart, kickstart and preseed are currently supported.
+  def pxe_type
+    "ZTP"
+  end
+
+  def available_loaders
+    ["None"]
```
Comment (ohadlevy) [inline app/models/operatingsystems/vrp.rb:19]: is that purely a UI thing? it might be better to return an empty array instead?

Comment (UXabre) [reply app/models/operatingsystems/vrp.rb:19]: Hey @ohadlevy , I actually used the code from JunOS as inspiration in which such code is also used. If you like, I can make it an empty array but I'll have to validate any implications this might have (at least, from my end, I don't use this at all)

File: app/views/unattended/provisioning_templates/ZTP/vrp_default_ztp_config.erb (hunk 3 above)
Comment (UXabre) [reply app/views/unattended/provisioning_templates/ZTP/vrp_default_ztp_config.erb]: Hmm, perhaps I've also misinterpreted that comment but, could it be that I **also** have to add this to community templates (so basically copy it there for others to be continue working on it?).

Also, I just noticed that I have used Base64.decode which is considered unsafe. I'm looking with Huawei how I can use a SHA265 hash instead but their documentation isn't too clear on what they did regarding the encoding (feels like some base32 algorithm but with a differnt dictionary). I don't think a template requiring unsafe methods is allowed at least?

Comment (timogoebel) [reply app/views/unattended/provisioning_templates/ZTP/vrp_default_ztp_config.erb]: The community templates are the main place the templates for Foreman live. We copy them to this repo before every release. So please add the template both here and at the community templates repo for now.

Comment (UXabre) [reply app/views/unattended/provisioning_templates/ZTP/vrp_default_ztp_config.erb]: Aaah, okay, clarifies a lot of things thank you @timogoebel, will do so

File: app/models/operatingsystems/vrp.rb (hunk 8 above)
Comment (UXabre) [reply app/models/operatingsystems/vrp.rb:19]: Hi @ohadlevy , just gave it a spin with an empty list of loaders, however, it doesn't work. Looking more closely, it seems the system checks to see if a pxe_loader has been selected:
https://github.com/theforeman/foreman/blob/develop/app/models/concerns/orchestration/tftp.rb#L18

I also noticed that NXOS & Solaris do it like that, probably for the same reason.

IMHO if we want to tackle this, this should be a different issue which deals with the implications of altering that check, no? Unless you see a harmless way forward?

**Comment by UXabre** [comment]
I have patched this code based upon your input. The only thing I'm addressing with Huawei is the correct encoding of the password in SHA-256 which is the reason why it has taken a bit longer to finalize this PR. They told me I should expect and answer by the 20th november at the latest. In my current template I have to decode a base64 type of password but this is an unsafe template code, so would probably make it "unusable" for most users

File: app/models/operatingsystems/junos.rb (hunk 7 above)
Comment (UXabre) [reply app/models/operatingsystems/junos.rb]: Removed the ".cfg"-suffix from JunOS :-)

**Comment by UXabre** [comment]
Okay, so I just got word back from Huawei, they claim there is no mapping from SHA256 to their format, which is quite annoying to say the least. Therefor I'd have to revert back to using Base64.decode in our template.

**On a side note:** is there any reason Base64.Decode is considered unsafe in Foreman (besides the fact that it returns a plaintext password)?

Will push the changes tonight so it can be further evaluated

File: app/models/operatingsystems/vrp.rb (hunk 8 above)
Comment (lzap) [reply app/models/operatingsystems/vrp.rb:19]: This was designed that way - PXELoader is a string and "None" is one of the allowed options, we check for this value specifically. Looks good to me.

**Comment by lzap** [comment]
Drop a comment once you get there, thanks.

**Comment by UXabre** [comment]
Hello @lzap , just pushed the requested changes. Sorry for it being a bit later.

**Comment by UXabre** [comment]
Just checked and the build fails because currently it still uses Base64. Will remove the template for now. My new contact within Huawei is quite helpful in helping me understand their password hash method. So, either I patch this PR with an addition password method or I create a new PR for that.

**Comment by UXabre** [comment]
Removed the template for now, should build now

File: lib/net/dhcp/ztp_record.rb [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+      self.vendor      = opts[:vendor]
+    end
+
+    def attrs
+      super.merge(
+        {
+          :ztp_firmware   => firmware,
+          :ztp_vendor     => vendor
+        }).compact
+    end
```
Comment (lzap) [inline lib/net/dhcp/ztp_record.rb:17]: Can you use Ruby memorization here and change this the same way for all record types? We call `attrs` method multiple times during figuring out dupes and this is slow.

**Review by lzap — CHANGES_REQUESTED** [review]
Looks good, one minor comment. Let me know when this is ready for testing, I will just test for regressions in `default_pxe_render` method and this can go into develop branch. Make sure it works with the real hardware as I have no chance testing this :-)

File: lib/net/dhcp/ztp_record.rb (hunk 9 above)
Comment (UXabre) [reply lib/net/dhcp/ztp_record.rb:17]: Hey @lzap , great input! Fixed it and also used "compact" for the other record types.

**Review by lzap — APPROVED** [review]
Thanks is this ready for testing? All I can do is rendering the templates as I don't have the hardware.

**Comment by UXabre** [comment]
Hi @lzap , sorry, I wasn't clear that this was ready for testing. Tested this in my set-up and works to expectation. Let me know if you need anything else :-)

**Review by lzap — APPROVED** [review]
Thanks.

**Comment by lzap** [comment]
Two failures unrelated. Merging.

**Comment by UXabre** [comment]
Hi @lzap, I've also resolved the conflict for the companion-PR for the smart-proxy: https://github.com/theforeman/smart-proxy/pull/622
